File: ios/ReactNativeSignaturesModule.swift
import ExpoModulesCore
import UIKit

struct SignatureSaveOptions: Record {
  @Field var format: String = "png"
  @Field var quality: Double = 0.92
  /** ARGB int produced by React Native's processColor. */
  @Field var backgroundColor: Int?
  @Field var trim: Bool = false
}

private func color(fromARGB value: Int?) -> UIColor? {
  guard let value else {
    return nil
  }
  let argb = UInt32(truncatingIfNeeded: value)
  return UIColor(
    red: CGFloat((argb >> 16) & 0xFF) / 255.0,
    green: CGFloat((argb >> 8) & 0xFF) / 255.0,
    blue: CGFloat(argb & 0xFF) / 255.0,
    alpha: CGFloat((argb >> 24) & 0xFF) / 255.0
  )
}

public class ReactNativeSignaturesModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ReactNativeSignatures")

    View(SignatureView.self) {
      Events("onBegin", "onEnd", "onInkChange")

      Prop("penColor") { (view: SignatureView, value: Int?) in
        view.penColor = color(fromARGB: value) ?? .black
      }

      Prop("minStrokeWidth") { (view: SignatureView, value: Double?) in
        view.minWidth = CGFloat(value ?? 0.5)
      }

      Prop("maxStrokeWidth") { (view: SignatureView, value: Double?) in
        view.maxWidth = CGFloat(value ?? 2.5)
      }

      Prop("velocityFilterWeight") { (view: SignatureView, value: Double?) in
        view.velocityFilterWeight = CGFloat(value ?? 0.7)
      }

      Prop("minDistance") { (view: SignatureView, value: Double?) in
        view.minDistance = CGFloat(value ?? 5)
      }

      Prop("dotSize") { (view: SignatureView, value: Double?) in
        view.dotSize = CGFloat(value ?? 0)
      }

      AsyncFunction("clear") { (view: SignatureView, promise: Promise) in
        DispatchQueue.main.async {
          view.clear()
          promise.resolve()
        }
      }

      AsyncFunction("isEmpty") { (view: SignatureView, promise: Promise) in
        DispatchQueue.main.async {
          promise.resolve(view.isEmptySignature)
        }
      }

      AsyncFunction("save") { (view: SignatureView, options: SignatureSaveOptions, promise: Promise) in
        DispatchQueue.main.async {
          let isJpeg = options.format == "jpeg"
          let background = color(fromARGB: options.backgroundColor) ?? (isJpeg ? UIColor.white : nil)
          guard let image = view.renderExportImage(trim: options.trim, backgroundColor: background) else {
            promise.reject("ERR_SIGNATURE_SAVE", "The signature surface is not ready yet")
            return
          }
          let quality = min(max(options.quality, 0), 1)
          DispatchQueue.global(qos: .userInitiated).async {
            let uiImage = UIImage(cgImage: image)
            let data = isJpeg ? uiImage.jpegData(compressionQuality: quality) : uiImage.pngData()
            if let data {
              promise.resolve(data.base64EncodedString())
            } else {
              promise.reject("ERR_SIGNATURE_SAVE", "Failed to encode the signature image")
            }
          }
        }
      }
    }
  }
}

File: android/src/main/java/expo/modules/signatures/ReactNativeSignaturesModule.kt
package expo.modules.signatures

import android.graphics.Bitmap
import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.util.Base64
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.io.ByteArrayOutputStream

class SignatureSaveOptions : Record {
  @Field val format: String = "png"
  @Field val quality: Double = 0.92
  /** ARGB int produced by React Native's processColor. */
  @Field val backgroundColor: Int? = null
  @Field val trim: Boolean = false
}

class ReactNativeSignaturesModule : Module() {
  private val mainHandler = Handler(Looper.getMainLooper())

  override fun definition() = ModuleDefinition {
    Name("ReactNativeSignatures")

    View(SignatureView::class) {
      Events("onBegin", "onEnd", "onInkChange")

      Prop("penColor") { view: SignatureView, value: Int? ->
        view.penColor = value ?: Color.BLACK
      }

      // Deliberately NOT named minWidth/maxWidth: those are reserved Yoga
      // layout prop names and would clamp the view's own size on Fabric.
      Prop("minStrokeWidth") { view: SignatureView, value: Double? ->
        view.minWidth = value ?: 0.5
      }

      Prop("maxStrokeWidth") { view: SignatureView, value: Double? ->
        view.maxWidth = value ?: 2.5
      }

      Prop("velocityFilterWeight") { view: SignatureView, value: Double? ->
        view.velocityFilterWeight = value ?: 0.7
      }

      Prop("minDistance") { view: SignatureView, value: Double? ->
        view.minDistance = value ?: 5.0
      }

      Prop("dotSize") { view: SignatureView, value: Double? ->
        view.dotSize = value ?: 0.0
      }

      AsyncFunction("clear") { view: SignatureView, promise: Promise ->
        mainHandler.post {
          view.clear()
          promise.resolve(null)
        }
      }

      AsyncFunction("isEmpty") { view: SignatureView, promise: Promise ->
        mainHandler.post {
          promise.resolve(view.isEmptySignature)
        }
      }

      AsyncFunction("save") { view: SignatureView, options: SignatureSaveOptions, promise: Promise ->
        mainHandler.post {
          val isJpeg = options.format == "jpeg"
          val background = options.backgroundColor ?: if (isJpeg) Color.WHITE else null
          val snapshot = view.renderExportBitmap(options.trim, background)
          if (snapshot == null) {
            promise.reject("ERR_SIGNATURE_SAVE", "The signature surface is not ready yet", null)
            return@post
          }
          Thread {
            try {
              val stream = ByteArrayOutputStream()
              val quality = (options.quality.coerceIn(0.0, 1.0) * 100).toInt()
              val format = if (isJpeg) Bitmap.CompressFormat.JPEG else Bitmap.CompressFormat.PNG
              snapshot.compress(format, quality, stream)
              promise.resolve(Base64.encodeToString(stream.toByteArray(), Base64.NO_WRAP))
            } catch (e: Throwable) {
              promise.reject("ERR_SIGNATURE_SAVE", "Failed to encode the signature image", e)
            }
          }.start()
        }
      }
    }
  }
}
